{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search datasets on Zenodo\n",
    "Ref: https://help.zenodo.org/guides/search/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def search_zenodo_datasets(query, size):\n",
    "    base_url = \"https://zenodo.org/api/records/\"\n",
    "    params = {\n",
    "        \"q\": query,\n",
    "        \"size\": size,  # Number of results to retrieve, adjust as needed\n",
    "        \"type\": \"dataset\", \n",
    "    }\n",
    "\n",
    "    try:\n",
    "        response = requests.get(base_url, params=params)\n",
    "        response.raise_for_status()  # Check for HTTP errors\n",
    "\n",
    "        # Parse the JSON response\n",
    "        data = response.json()\n",
    "        hits = data.get(\"hits\", {}).get(\"hits\", [])\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        print(\"Error:\", e)\n",
    "\n",
    "    return hits\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    query = \"citeseer\"  # Replace with your search query\n",
    "    hits = search_zenodo_datasets(query, 10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'conceptdoi': '10.5281/zenodo.8226021',\n",
       "  'conceptrecid': '8226021',\n",
       "  'created': '2023-08-08T19:02:08.004663+00:00',\n",
       "  'doi': '10.5281/zenodo.8226022',\n",
       "  'files': [{'bucket': '2c35258e-4aa4-468f-99c8-cb5af2eb856d',\n",
       "    'checksum': 'md5:223f29affb86dce9e08deb586e91a4e8',\n",
       "    'key': 'dataset.tar.gz',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/2c35258e-4aa4-468f-99c8-cb5af2eb856d/dataset.tar.gz'},\n",
       "    'size': 2665450375,\n",
       "    'type': 'gz'},\n",
       "   {'bucket': '2c35258e-4aa4-468f-99c8-cb5af2eb856d',\n",
       "    'checksum': 'md5:88089cf5d64e0e0f5bbb7ee70a81fcfd',\n",
       "    'key': 'read_data.ipynb',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/2c35258e-4aa4-468f-99c8-cb5af2eb856d/read_data.ipynb'},\n",
       "    'size': 1672,\n",
       "    'type': 'ipynb'}],\n",
       "  'id': 8226022,\n",
       "  'links': {'badge': 'https://zenodo.org/badge/doi/10.5281/zenodo.8226022.svg',\n",
       "   'bucket': 'https://zenodo.org/api/files/2c35258e-4aa4-468f-99c8-cb5af2eb856d',\n",
       "   'conceptbadge': 'https://zenodo.org/badge/doi/10.5281/zenodo.8226021.svg',\n",
       "   'conceptdoi': 'https://doi.org/10.5281/zenodo.8226021',\n",
       "   'doi': 'https://doi.org/10.5281/zenodo.8226022',\n",
       "   'html': 'https://zenodo.org/record/8226022',\n",
       "   'latest': 'https://zenodo.org/api/records/8226022',\n",
       "   'latest_html': 'https://zenodo.org/record/8226022',\n",
       "   'self': 'https://zenodo.org/api/records/8226022'},\n",
       "  'metadata': {'access_right': 'open',\n",
       "   'access_right_category': 'success',\n",
       "   'creators': [{'affiliation': 'Michigan State University',\n",
       "     'name': 'Juanhui Li',\n",
       "     'orcid': '0000-0003-4909-1778'},\n",
       "    {'affiliation': 'Michigan State University', 'name': 'Harry Shomer'},\n",
       "    {'affiliation': 'Michigan State University', 'name': 'Haitao Mao'},\n",
       "    {'affiliation': 'Michigan State University', 'name': 'Shenglai Zeng'},\n",
       "    {'affiliation': 'New Jersey Institute of Technology', 'name': 'Yao Ma'},\n",
       "    {'affiliation': 'Snap Inc.', 'name': 'Neil Shah'},\n",
       "    {'affiliation': 'Michigan State University', 'name': 'Jiliang Tang'},\n",
       "    {'affiliation': 'Baidu Inc.', 'name': 'Dawei Yin'}],\n",
       "   'description': '<p>The supplementary dataset for the paper &quot;Evaluating Graph Neural Networks for Link Prediction: Current Pitfalls and New Benchmarking&quot;. We include the splits for cora, citeseer, and pubmed, the hard negative samples, and the node2vec embeddings. We also include a jupyter file <em>read_data.ipynb</em> to show how to read the non-txt file.</p>\\n\\n<ul>\\n\\t<li>heart_test_samples.npy,&nbsp;heart_valid_samples.npy: the heard negative samples</li>\\n\\t<li>*-n2v-embedding.pt: node2vec embeddings</li>\\n\\t<li>test_samples_index.pt, valid_samples_index.pt: the node index of the selected samples in ogbl-ppa under HeaRT</li>\\n\\t<li>gnn_feature: the input feature of cora, citeseer, pubmed</li>\\n</ul>\\n\\n<p>More details for our code&nbsp;and how to use the dataset are&nbsp;on the code repository:&nbsp;https://github.com/Juanhui28/HeaRT .</p>',\n",
       "   'doi': '10.5281/zenodo.8226022',\n",
       "   'keywords': ['link prediction, graph neural networks, benchmark, hard negative samples'],\n",
       "   'license': {'id': 'CC-BY-4.0'},\n",
       "   'publication_date': '2023-08-08',\n",
       "   'related_identifiers': [{'identifier': '10.5281/zenodo.8226021',\n",
       "     'relation': 'isVersionOf',\n",
       "     'scheme': 'doi'}],\n",
       "   'relations': {'version': [{'count': 1,\n",
       "      'index': 0,\n",
       "      'is_last': True,\n",
       "      'last_child': {'pid_type': 'recid', 'pid_value': '8226022'},\n",
       "      'parent': {'pid_type': 'recid', 'pid_value': '8226021'}}]},\n",
       "   'resource_type': {'title': 'Dataset', 'type': 'dataset'},\n",
       "   'title': 'Supplementary dataset for \"Evaluating Graph Neural Networks for Link Prediction: Current Pitfalls and New Benchmarking\"',\n",
       "   'version': 'v1'},\n",
       "  'owners': [594797],\n",
       "  'revision': 2,\n",
       "  'stats': {'downloads': 1.0,\n",
       "   'unique_downloads': 1.0,\n",
       "   'unique_views': 10.0,\n",
       "   'version_downloads': 1.0,\n",
       "   'version_unique_downloads': 1.0,\n",
       "   'version_unique_views': 10.0,\n",
       "   'version_views': 12.0,\n",
       "   'version_volume': 1672.0,\n",
       "   'views': 12.0,\n",
       "   'volume': 1672.0},\n",
       "  'updated': '2023-08-09T02:26:49.823489+00:00'},\n",
       " {'conceptdoi': '10.5281/zenodo.5564161',\n",
       "  'conceptrecid': '5564161',\n",
       "  'created': '2022-11-10T17:13:03.219236+00:00',\n",
       "  'doi': '10.5281/zenodo.7311418',\n",
       "  'files': [{'bucket': '564bd3e4-0245-4d4c-bcae-541089d4ccef',\n",
       "    'checksum': 'md5:3458912f4358ef481fab8815a6759d66',\n",
       "    'key': 'data.uniq.fasta',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/564bd3e4-0245-4d4c-bcae-541089d4ccef/data.uniq.fasta'},\n",
       "    'size': 1163414269,\n",
       "    'type': 'fasta'},\n",
       "   {'bucket': '564bd3e4-0245-4d4c-bcae-541089d4ccef',\n",
       "    'checksum': 'md5:2d8e993fc6a13297858f09e22cf4b906',\n",
       "    'key': 'eden_plant_postclean_100sim.rds',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/564bd3e4-0245-4d4c-bcae-541089d4ccef/eden_plant_postclean_100sim.rds'},\n",
       "    'size': 201257,\n",
       "    'type': 'rds'},\n",
       "   {'bucket': '564bd3e4-0245-4d4c-bcae-541089d4ccef',\n",
       "    'checksum': 'md5:2b7a36e9acdbd4ffd7a63b02e1917e8d',\n",
       "    'key': 'metabaR_Eden_Plants_100sim.html',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/564bd3e4-0245-4d4c-bcae-541089d4ccef/metabaR_Eden_Plants_100sim.html'},\n",
       "    'size': 2923537,\n",
       "    'type': 'html'},\n",
       "   {'bucket': '564bd3e4-0245-4d4c-bcae-541089d4ccef',\n",
       "    'checksum': 'md5:59201b2d1f44471c36dc575f5f7a981f',\n",
       "    'key': 'ngsfilter_TRNL_PLANTS_EDEN_PROJECTb.txt',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/564bd3e4-0245-4d4c-bcae-541089d4ccef/ngsfilter_TRNL_PLANTS_EDEN_PROJECTb.txt'},\n",
       "    'size': 285109,\n",
       "    'type': 'txt'}],\n",
       "  'id': 7311418,\n",
       "  'links': {'badge': 'https://zenodo.org/badge/doi/10.5281/zenodo.7311418.svg',\n",
       "   'bucket': 'https://zenodo.org/api/files/564bd3e4-0245-4d4c-bcae-541089d4ccef',\n",
       "   'conceptbadge': 'https://zenodo.org/badge/doi/10.5281/zenodo.5564161.svg',\n",
       "   'conceptdoi': 'https://doi.org/10.5281/zenodo.5564161',\n",
       "   'doi': 'https://doi.org/10.5281/zenodo.7311418',\n",
       "   'html': 'https://zenodo.org/record/7311418',\n",
       "   'latest': 'https://zenodo.org/api/records/7311418',\n",
       "   'latest_html': 'https://zenodo.org/record/7311418',\n",
       "   'self': 'https://zenodo.org/api/records/7311418'},\n",
       "  'metadata': {'access_right': 'open',\n",
       "   'access_right_category': 'success',\n",
       "   'creators': [{'affiliation': 'Centre for Ecology and Conservation, University of Exeter, Penryn, UK',\n",
       "     'name': 'Donald, Julian',\n",
       "     'orcid': '0000-0001-6900-3777'},\n",
       "    {'affiliation': 'Evolution et Diversité Biologique (EDB UMR 5174), Université Toulouse 3 Paul Sabatier',\n",
       "     'name': 'Iribar, Amaia',\n",
       "     'orcid': '0000-0002-4709-648X'},\n",
       "    {'affiliation': 'Evolution et Diversité Biologique (EDB UMR 5174), Université Toulouse 3 Paul Sabatier',\n",
       "     'name': 'Chave, Jerome',\n",
       "     'orcid': '0000-0002-7766-1347'}],\n",
       "   'description': '<p>The project seeks to understand the potential for the use of eDNA collected from soil to characterise plant communities. To do so, soils were sampled at the Eden Project in Cornwall UK, within the two covered biomes where we have a good understanding of the structure and composition of plant communities (further quantified with above ground plant coverage inventories).&nbsp;32 plots were established across 10 different plant assemblages, each of which experiences subtle differences in soil chemistry and microclimate. Each plot consists of a 2 x 2 m quadrat, with four soil aggregates collected at each corner.&nbsp;</p>\\n\\n<p>eDNA was then extracted and amplified following the methods detailed in Zinger et al. (2016) and Donald et al. (2021). The primers used targeted the P6 loop of thechloroplastic &nbsp;trnL intron [primer_fwd: GGGCAATCCTGAGCCAA, primer_rev: CCATTGAGTCTCTGCACCTATC] (Taberlet et al. 2007). 16 Extraction, 54 Sequencing, and 16 PCR controls are included so as to account for potential errors generated during the processing of samples, with a mock community (4 positive controls) of 10 known plant sequences also included to guide filtering thresholds.&nbsp;PCR products were pooled and sequencing libraries were constructed using the Illumina TruSeq NanoPCRFree kit following the supplier&rsquo;s instructions (Illumina Inc., San Diego, California, USA), except that the ligation product was not PCR amplified to limit tag-jump biases (Taberlet et al 2018). The libraries were then sequenced on an Illumina Hiseq platform&nbsp;(San Diego, CA, USA).</p>\\n\\n<p>Sequencing was conducted by the GenoToul bioinformatics platform (Toulouse, France), with the OBITOOLS package (Boyer et al. 2016). Here, the produced sequence data was processed using the following steps. First, &lsquo;illuminapairedend&rsquo; was used to assemble paired-end reads. This algorithm is based on an exact alignment algorithm that considers the quality scores at all positions during the assembly process. Subsequently, we used the &lsquo;ngsfilter&rsquo; command to identify and remove the primers and tags on each read, and assign reads to their respective samples (NGS filter file provided: <strong>ngsfilter_TRNL_PLANTS_EDEN_PROJECTb.txt</strong>). This program was used with its default parameters tolerating two mismatches for each of the two primers and no mismatch for the tags. Following this, sequencing reads were dereplicated using the &lsquo;obiuniq&rsquo; command. The produced <strong>data.uniq.fasta</strong> file is supplied here. Sequences were then further filtered to remove sequences of low quality (containing Ns or with paired-end alignment scores below 50), and sequences represented by only one read (singletons)&nbsp;using the &lsquo;obigrep&rsquo; command. To remove PCR/sequencing errors as well as intraspecific variability, we built OTUs (Operational Taxonomic Units) using the &lsquo;sumaclust&rsquo; clustering algorithm (Mercier et al. 2013), which considers the most abundant sequence of each cluster as the cluster representative. &nbsp;OTUs were set at a sequence similarity threshold of&nbsp;95%. To assign a taxon to plant&nbsp;OTUs, we built a&nbsp;reference sequence database using the ecoPCR programme (Ficetola et al. 2010) on the European Molecular Biology Laboratory (EMBL; release 141).&nbsp;OTUs were then assigned a taxonomy, using OBITOOL&rsquo;s ecotag programme (Boyer et al. 2016), which performs a global alignment of each OTU sequence (the query) against each reference. The reference taxon assigned to each OTU corresponds to the Last Common Ancestor of all the best-match sequences for the query.&nbsp;</p>\\n\\n<p><br>\\nDatasets were subsequently filtered to remove contaminants as well as artefacts such as PCR chimeras and remaining sequencing errors, using routines implemented in the metabaR R package (Zinger et al 2021), in R version 3.6.1 (R Development Core Team, 2013).&nbsp;The filtering process consisted of four steps: (i) a negative control-based filtering. OTUs whose maximum abundance was found in extraction/PCR negative controls were removed from the dataset, as they were likely to be reagent/aerosol contaminants, better amplified in the absence of competing DNA fragments as it is the case in biological samples. (ii) a reference-based filtering. OTUs which are too dissimilar from sequences available in reference databases are potential chimeras generated during sequencing and amplification. In this study, we chose to set similarity thresholds at 100%. (iii) an abundance-based filtering. This procedure targets incorrect assignment of a few numbers of sequences corresponding to true OTUs occurring to the wrong sample, a phenomenon called &ldquo;tag-switching&rdquo;. It consists in setting OTUs abundances to 0 in samples where their abundance represents &lt; 0.03% of the total OTU abundance in the entire dataset. (iv) Finally, we conducted a PCR-based filtering by considering any PCR reaction that yielded less than 1000 reads&nbsp;as non-functional, and removed them from the dataset. The script used for implementing this is provided (<strong>metabaR_Eden_Plants_100sim.html)</strong>, with sequence data processed to remove contaminants, OTUs of low taxonomic resolution, and PCRs with too low a read count. The clean data is provided (<strong>eden_plant_postclean_100sim.rds</strong>).</p>\\n\\n<p>References:</p>\\n\\n<p>Boyer, F. <em>et al.</em> (2016) &lsquo;obitools: a unix-inspired software package for DNA metabarcoding&rsquo;, <em>Molecular Ecology Resources</em>, 16(1), pp. 176&ndash;182. doi:<a href=\"https://doi.org/10.1111/1755-0998.12428\">10.1111/1755-0998.12428</a>.</p>\\n\\n<p>Donald, J.&nbsp;<em>et al. (2021) &#39;</em>&lsquo;Multi-taxa environmental DNA inventories reveal distinct taxonomic and functional diversity in urban tropical forest fragments.&lsquo;&nbsp;<em>Global Ecology and Conservation</em>&nbsp;29 (2021): e01724.</p>\\n\\n<p>Mercier, C. <em>et al.</em> (2013) &lsquo;SUMATRA and SUMACLUST: fast and exact comparison and clustering of sequences&rsquo;, in <em>Programs and Abstracts of the SeqBio 2013 workshop. Abstract</em>. Citeseer, pp. 27&ndash;29.</p>\\n\\n<p>Taberlet, P. <em>et al.</em> (2007) &lsquo;Power and limitations of the chloroplast trn L (UAA) intron for plant DNA barcoding&rsquo;, <em>Nucleic Acids Research</em>, 35(3), pp. e14&ndash;e14. doi:<a href=\"https://doi.org/10.1093/nar/gkl938\">10.1093/nar/gkl938</a>.</p>\\n\\n<p>Taberlet, P. <em>et al.</em> (2018) <em>Environmental DNA: For Biodiversity Research and Monitoring</em>. Oxford University Press.</p>\\n\\n<p>Team, R.C. (2013) <em>R: A language and environment for statistical computing</em>. Vienna, Austria.</p>\\n\\n<p>Zinger, L. <em>et al.</em> (2016) &lsquo;Extracellular DNA extraction is a fast, cheap and reliable alternative for multi-taxa surveys based on soil DNA&rsquo;, <em>Soil Biology and Biochemistry</em>, 96, pp. 16&ndash;19.</p>\\n\\n<p>Zinger, L. et al. (2021) &lsquo;metabaR: An r package for the evaluation and improvement of DNA metabarcoding data quality&rsquo;, Methods in Ecology and Evolution. DOI:&nbsp;<a href=\"https://doi.org/10.1111/2041-210X.13552\">https://doi.org/10.1111/2041-210X.13552</a></p>',\n",
       "   'doi': '10.5281/zenodo.7311418',\n",
       "   'grants': [{'acronym': 'MICOCO',\n",
       "     'code': '101024135',\n",
       "     'funder': {'acronyms': [],\n",
       "      'doi': '10.13039/501100000780',\n",
       "      'links': {'self': 'https://zenodo.org/api/funders/10.13039/501100000780'},\n",
       "      'name': 'European Commission'},\n",
       "     'links': {'self': 'https://zenodo.org/api/grants/10.13039/501100000780::101024135'},\n",
       "     'program': 'H2020',\n",
       "     'title': 'How do biotic and environmental variation affect soil microbial community composition and functioning across spatial scales?'},\n",
       "    {'acronym': 'CEBA',\n",
       "     'code': '10-LABX-0025',\n",
       "     'funder': {'acronyms': ['Not Available'],\n",
       "      'doi': '10.13039/501100001665',\n",
       "      'links': {'self': 'https://zenodo.org/api/funders/10.13039/501100001665'},\n",
       "      'name': 'Agence Nationale de la Recherche'},\n",
       "     'links': {'self': 'https://zenodo.org/api/grants/10.13039/501100001665::10-LABX-0025'},\n",
       "     'program': '',\n",
       "     'title': 'CEnter of the study of Biodiversity in Amazonia'},\n",
       "    {'acronym': 'TULIP',\n",
       "     'code': '10-LABX-0041',\n",
       "     'funder': {'acronyms': ['Not Available'],\n",
       "      'doi': '10.13039/501100001665',\n",
       "      'links': {'self': 'https://zenodo.org/api/funders/10.13039/501100001665'},\n",
       "      'name': 'Agence Nationale de la Recherche'},\n",
       "     'links': {'self': 'https://zenodo.org/api/grants/10.13039/501100001665::10-LABX-0041'},\n",
       "     'program': '',\n",
       "     'title': 'Towards a Unified theory of biotic Interactions: the roLe of environmental'}],\n",
       "   'license': {'id': 'CC-BY-4.0'},\n",
       "   'publication_date': '2021-10-12',\n",
       "   'related_identifiers': [{'identifier': '10.5281/zenodo.5564161',\n",
       "     'relation': 'isVersionOf',\n",
       "     'scheme': 'doi'}],\n",
       "   'relations': {'version': [{'count': 2,\n",
       "      'index': 1,\n",
       "      'is_last': True,\n",
       "      'last_child': {'pid_type': 'recid', 'pid_value': '7311418'},\n",
       "      'parent': {'pid_type': 'recid', 'pid_value': '5564161'}}]},\n",
       "   'resource_type': {'title': 'Dataset', 'type': 'dataset'},\n",
       "   'title': 'Plant Communities at the Eden Project, UK, derived from soil eDNA'},\n",
       "  'owners': [240243],\n",
       "  'revision': 2,\n",
       "  'stats': {'downloads': 3.0,\n",
       "   'unique_downloads': 2.0,\n",
       "   'unique_views': 10.0,\n",
       "   'version_downloads': 25.0,\n",
       "   'version_unique_downloads': 9.0,\n",
       "   'version_unique_views': 31.0,\n",
       "   'version_views': 35.0,\n",
       "   'version_volume': 2368699911.0,\n",
       "   'views': 11.0,\n",
       "   'volume': 3493755.0},\n",
       "  'updated': '2022-11-11T02:26:33.521500+00:00'},\n",
       " {'conceptdoi': '10.5281/zenodo.3732152',\n",
       "  'conceptrecid': '3732152',\n",
       "  'created': '2020-03-28T20:15:17.500346+00:00',\n",
       "  'doi': '10.5281/zenodo.3732153',\n",
       "  'files': [{'bucket': '65aa6283-4db9-4a43-abba-2c9907ffb8aa',\n",
       "    'checksum': 'md5:d86909436e70af09c529dbdd3512bf1e',\n",
       "    'key': 'attrCiteseer.csv.tar.bz2',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/65aa6283-4db9-4a43-abba-2c9907ffb8aa/attrCiteseer.csv.tar.bz2'},\n",
       "    'size': 23464612,\n",
       "    'type': 'bz2'},\n",
       "   {'bucket': '65aa6283-4db9-4a43-abba-2c9907ffb8aa',\n",
       "    'checksum': 'md5:a7a6374457c8576ed0adbeaeaf741cac',\n",
       "    'key': 'attrLastFm.csv.tar.bz2',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/65aa6283-4db9-4a43-abba-2c9907ffb8aa/attrLastFm.csv.tar.bz2'},\n",
       "    'size': 427399715,\n",
       "    'type': 'bz2'},\n",
       "   {'bucket': '65aa6283-4db9-4a43-abba-2c9907ffb8aa',\n",
       "    'checksum': 'md5:b9c6ff802f172aee31633d09cea2ffc7',\n",
       "    'key': 'graphCiteseer.csv.tar.bz2',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/65aa6283-4db9-4a43-abba-2c9907ffb8aa/graphCiteseer.csv.tar.bz2'},\n",
       "    'size': 23193022,\n",
       "    'type': 'bz2'},\n",
       "   {'bucket': '65aa6283-4db9-4a43-abba-2c9907ffb8aa',\n",
       "    'checksum': 'md5:06e1513c274f989dea179111d7d1d73f',\n",
       "    'key': 'graphLastFm.csv.tar.gz',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/65aa6283-4db9-4a43-abba-2c9907ffb8aa/graphLastFm.csv.tar.gz'},\n",
       "    'size': 2650018,\n",
       "    'type': 'gz'},\n",
       "   {'bucket': '65aa6283-4db9-4a43-abba-2c9907ffb8aa',\n",
       "    'checksum': 'md5:619cf8400c93db11c3f819ae92844834',\n",
       "    'key': 'newAttrDBLP.csv.tar.bz2',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/65aa6283-4db9-4a43-abba-2c9907ffb8aa/newAttrDBLP.csv.tar.bz2'},\n",
       "    'size': 2815720,\n",
       "    'type': 'bz2'},\n",
       "   {'bucket': '65aa6283-4db9-4a43-abba-2c9907ffb8aa',\n",
       "    'checksum': 'md5:550c3cc57f6b1ab118328a603ab5132b',\n",
       "    'key': 'newGraphDBLP.csv.tar.bz2',\n",
       "    'links': {'self': 'https://zenodo.org/api/files/65aa6283-4db9-4a43-abba-2c9907ffb8aa/newGraphDBLP.csv.tar.bz2'},\n",
       "    'size': 3202441,\n",
       "    'type': 'bz2'}],\n",
       "  'id': 3732153,\n",
       "  'links': {'badge': 'https://zenodo.org/badge/doi/10.5281/zenodo.3732153.svg',\n",
       "   'bucket': 'https://zenodo.org/api/files/65aa6283-4db9-4a43-abba-2c9907ffb8aa',\n",
       "   'conceptbadge': 'https://zenodo.org/badge/doi/10.5281/zenodo.3732152.svg',\n",
       "   'conceptdoi': 'https://doi.org/10.5281/zenodo.3732152',\n",
       "   'doi': 'https://doi.org/10.5281/zenodo.3732153',\n",
       "   'html': 'https://zenodo.org/record/3732153',\n",
       "   'latest': 'https://zenodo.org/api/records/3732153',\n",
       "   'latest_html': 'https://zenodo.org/record/3732153',\n",
       "   'self': 'https://zenodo.org/api/records/3732153'},\n",
       "  'metadata': {'access_right': 'open',\n",
       "   'access_right_category': 'success',\n",
       "   'contributors': [],\n",
       "   'creators': [{'affiliation': 'UCSB', 'name': 'Silva, Arlei'},\n",
       "    {'affiliation': 'UFMG', 'name': 'Meira, Jr. Wagner'},\n",
       "    {'affiliation': 'RPI',\n",
       "     'name': 'Zaki, Mohammed J.',\n",
       "     'orcid': '0000-0003-4711-0234'}],\n",
       "   'description': '<p>## SCPM: An implementation of an algorithm for structural correlation pattern mining.</p>\\n\\n<p>The structural correlation measures how a set of attributes induces dense subgraphs in an attributed graph. A structural correlation pattern is a dense subgraph induced by a particular attribute set. Structural correlation pattern mining is useful to analyze how different attribute sets are correlated to dense subgraphs in several real-life attributed graphs.</p>\\n\\n<p>**Relevant Publications**</p>\\n\\n<p>* Arlei Silva, Wagner Meira, Jr., and Mohammed J. Zaki. Structural correlation pattern mining for large graphs. In Proceedings of the Eighth Workshop on Mining and Learning with Graphs (MLG &#39;10).</p>\\n\\n<p>* Arlei Silva, Wagner Meira, Jr., and Mohammed J. Zaki. Mining Attribute-structure Correlated Patterns in Large Attributed Graphs. In Proceedings of the VLDB Endowment (PVLDB &#39;12).</p>\\n\\n<p>* Arlei Silva. Structural correlation pattern mining for large graphs. M.Sc Thesis, Computer Science Department, Universidade Federal de Minas Gerais, 2011.</p>\\n\\n<p>* Arlei Silva, Wagner Meira Jr. Structural correlation pattern mining for large graphs. Thesis and Dissertation Contest of the Brazilian Computer Society (CTD&#39;12).</p>\\n\\n<p><br>\\n## HOW TO</p>\\n\\n<p>cd to trunk and run make<br>\\nsee README in trunk</p>\\n\\n<p><br>\\n## Datasets:</p>\\n\\n<p>### Description:</p>\\n\\n<p>#### ATTRIBUTE FILE:</p>\\n\\n<p>Format: Lists the attributes of each vertex from the graph.</p>\\n\\n<p>&nbsp; &nbsp; &lt;VERTEX_ID&gt;,&lt;ATTRIBUTE_ID&gt;,&lt;ATTRIBUTE_ID&gt;...,&lt;ATTRIBUTE_ID&gt;</p>\\n\\n<p>&nbsp; &nbsp; &nbsp; &nbsp; Example:&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 1,A,C&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 2,A&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 3,A,C,D&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 4,A,D&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 5,A,E&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 6,A,B,C&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 7,A,B,E&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 8,A,B&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 9,A,B&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 10,A,B,D&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 11,A,B</p>\\n\\n<p>#### GRAPH FILE:</p>\\n\\n<p>Format: Lists the neighbors of each vertex from the graph (adjacency list). Although the graph is undirected, each edge must be included in both directions.</p>\\n\\n<p>&nbsp; &nbsp; &lt;VERTEX_ID&gt;,&lt;NEIGHBOR_ID&gt;,&lt;NEIGHBOR_ID&gt;...,&lt;NEIGHBOR_ID&gt;&nbsp;<br>\\n&nbsp; &nbsp;&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; Example:&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 1,4&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 2,3&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 3,2,4,5,6,7&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 4,1,3,5,6&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 5,3,4,6&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 6,3,4,5,7,8,9,10&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 7,3,6,8,11&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 8,6,7,9,10,11&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 9,6,8,10,11&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 10,6,8,9,11&nbsp;<br>\\n&nbsp; &nbsp; &nbsp; &nbsp; 11,7,8,9,10</p>\\n\\n<p>### REAL DATASETS</p>\\n\\n<p>Lastfm:</p>\\n\\n<p>attributes: attrLastFm.csv.tar.bz2</p>\\n\\n<p>network: graphLastFm.csv.tar.gz</p>\\n\\n<p>DBLP:</p>\\n\\n<p>attributes: newAttrDBLP.csv.tar.bz2</p>\\n\\n<p>network: newGraphDBLP.csv.tar.bz2</p>\\n\\n<p>CITESEER:</p>\\n\\n<p>attributes: attrCiteseer.csv.tar.bz2</p>\\n\\n<p>network: graphCiteseer.csv.tar.bz2</p>',\n",
       "   'doi': '10.5281/zenodo.3732153',\n",
       "   'grants': [{'acronym': '',\n",
       "     'code': '0829835',\n",
       "     'funder': {'acronyms': [],\n",
       "      'doi': '10.13039/100000001',\n",
       "      'links': {'self': 'https://zenodo.org/api/funders/10.13039/100000001'},\n",
       "      'name': 'National Science Foundation'},\n",
       "     'links': {'self': 'https://zenodo.org/api/grants/10.13039/100000001::0829835'},\n",
       "     'program': 'CISE/OAD',\n",
       "     'title': 'EMT/BSSE: Discovery of Gene and Protein Expression Patterns and Networks'}],\n",
       "   'journal': {'issue': '5',\n",
       "    'pages': '466--477',\n",
       "    'title': 'PVLDB',\n",
       "    'volume': '5'},\n",
       "   'keywords': ['graph mining, attribute-structure correlation pattern mining, social networks, attributed graphs'],\n",
       "   'license': {'id': 'CC-BY-4.0'},\n",
       "   'publication_date': '2020-03-28',\n",
       "   'references': ['Arlei Silva, Mohammed J. Zaki, and Wagner Meira Jr. Mining attribute-structure correlated patterns in large attributed graphs. PVLDB, 5(5):466–477, 2012.'],\n",
       "   'related_identifiers': [{'identifier': '10.5281/zenodo.3732152',\n",
       "     'relation': 'isVersionOf',\n",
       "     'scheme': 'doi'}],\n",
       "   'relations': {'version': [{'count': 1,\n",
       "      'index': 0,\n",
       "      'is_last': True,\n",
       "      'last_child': {'pid_type': 'recid', 'pid_value': '3732153'},\n",
       "      'parent': {'pid_type': 'recid', 'pid_value': '3732152'}}]},\n",
       "   'resource_type': {'title': 'Dataset', 'type': 'dataset'},\n",
       "   'thesis': {'supervisors': []},\n",
       "   'title': 'Graphs and Attributes used for the attribute-structure correlation pattern mining'},\n",
       "  'owners': [95836],\n",
       "  'revision': 2,\n",
       "  'stats': {'downloads': 59.0,\n",
       "   'unique_downloads': 29.0,\n",
       "   'unique_views': 112.0,\n",
       "   'version_downloads': 59.0,\n",
       "   'version_unique_downloads': 29.0,\n",
       "   'version_unique_views': 112.0,\n",
       "   'version_views': 120.0,\n",
       "   'version_volume': 4886215788.0,\n",
       "   'views': 120.0,\n",
       "   'volume': 4886215788.0},\n",
       "  'updated': '2020-03-29T20:20:13.136384+00:00'}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits\n",
    "        # if not hits:\n",
    "        #     print(\"No datasets found.\")\n",
    "        #     return\n",
    "\n",
    "        # for hit in hits:\n",
    "        #     record = hit.get(\"metadata\", {})\n",
    "        #     print(\"Title:\", record.get(\"title\", \"N/A\"))\n",
    "        #     print(\"DOI:\", record.get(\"doi\", \"N/A\"))\n",
    "        #     print(\"URL:\", record.get(\"url\", \"N/A\"))\n",
    "        #     print(\"Creators:\", \", \".join(record.get(\"creators\", [\"N/A\"])))\n",
    "        #     print(\"Description:\", record.get(\"description\", \"N/A\"))\n",
    "        #     print(\"=\"*30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"conceptdoi\": \"10.5281/zenodo.8226021\", \"conceptrecid\": \"8226021\", \"created\": \"2023-08-08T19:02:08.004663+00:00\", \"doi\": \"10.5281/zenodo.8226022\", \"files\": [{\"bucket\": \"2c35258e-4aa4-468f-99c8-cb5af2eb856d\", \"checksum\": \"md5:223f29affb86dce9e08deb586e91a4e8\", \"key\": \"dataset.tar.gz\", \"links\": {\"self\": \"https://zenodo.org/api/files/2c35258e-4aa4-468f-99c8-cb5af2eb856d/dataset.tar.gz\"}, \"size\": 2665450375, \"type\": \"gz\"}, {\"bucket\": \"2c35258e-4aa4-468f-99c8-cb5af2eb856d\", \"checksum\": \"md5:88089cf5d64e0e0f5bbb7ee70a81fcfd\", \"key\": \"read_data.ipynb\", \"links\": {\"self\": \"https://zenodo.org/api/files/2c35258e-4aa4-468f-99c8-cb5af2eb856d/read_data.ipynb\"}, \"size\": 1672, \"type\": \"ipynb\"}], \"id\": 8226022, \"links\": {\"badge\": \"https://zenodo.org/badge/doi/10.5281/zenodo.8226022.svg\", \"bucket\": \"https://zenodo.org/api/files/2c35258e-4aa4-468f-99c8-cb5af2eb856d\", \"conceptbadge\": \"https://zenodo.org/badge/doi/10.5281/zenodo.8226021.svg\", \"conceptdoi\": \"https://doi.org/10.5281/zenodo.8226021\", \"doi\": \"https://doi.org/10.5281/zenodo.8226022\", \"html\": \"https://zenodo.org/record/8226022\", \"latest\": \"https://zenodo.org/api/records/8226022\", \"latest_html\": \"https://zenodo.org/record/8226022\", \"self\": \"https://zenodo.org/api/records/8226022\"}, \"metadata\": {\"access_right\": \"open\", \"access_right_category\": \"success\", \"creators\": [{\"affiliation\": \"Michigan State University\", \"name\": \"Juanhui Li\", \"orcid\": \"0000-0003-4909-1778\"}, {\"affiliation\": \"Michigan State University\", \"name\": \"Harry Shomer\"}, {\"affiliation\": \"Michigan State University\", \"name\": \"Haitao Mao\"}, {\"affiliation\": \"Michigan State University\", \"name\": \"Shenglai Zeng\"}, {\"affiliation\": \"New Jersey Institute of Technology\", \"name\": \"Yao Ma\"}, {\"affiliation\": \"Snap Inc.\", \"name\": \"Neil Shah\"}, {\"affiliation\": \"Michigan State University\", \"name\": \"Jiliang Tang\"}, {\"affiliation\": \"Baidu Inc.\", \"name\": \"Dawei Yin\"}], \"description\": \"<p>The supplementary dataset for the paper &quot;Evaluating Graph Neural Networks for Link Prediction: Current Pitfalls and New Benchmarking&quot;. We include the splits for cora, citeseer, and pubmed, the hard negative samples, and the node2vec embeddings. We also include a jupyter file <em>read_data.ipynb</em> to show how to read the non-txt file.</p>\\n\\n<ul>\\n\\t<li>heart_test_samples.npy,&nbsp;heart_valid_samples.npy: the heard negative samples</li>\\n\\t<li>*-n2v-embedding.pt: node2vec embeddings</li>\\n\\t<li>test_samples_index.pt, valid_samples_index.pt: the node index of the selected samples in ogbl-ppa under HeaRT</li>\\n\\t<li>gnn_feature: the input feature of cora, citeseer, pubmed</li>\\n</ul>\\n\\n<p>More details for our code&nbsp;and how to use the dataset are&nbsp;on the code repository:&nbsp;https://github.com/Juanhui28/HeaRT .</p>\", \"doi\": \"10.5281/zenodo.8226022\", \"keywords\": [\"link prediction, graph neural networks, benchmark, hard negative samples\"], \"license\": {\"id\": \"CC-BY-4.0\"}, \"publication_date\": \"2023-08-08\", \"related_identifiers\": [{\"identifier\": \"10.5281/zenodo.8226021\", \"relation\": \"isVersionOf\", \"scheme\": \"doi\"}], \"relations\": {\"version\": [{\"count\": 1, \"index\": 0, \"is_last\": true, \"last_child\": {\"pid_type\": \"recid\", \"pid_value\": \"8226022\"}, \"parent\": {\"pid_type\": \"recid\", \"pid_value\": \"8226021\"}}]}, \"resource_type\": {\"title\": \"Dataset\", \"type\": \"dataset\"}, \"title\": \"Supplementary dataset for \\\"Evaluating Graph Neural Networks for Link Prediction: Current Pitfalls and New Benchmarking\\\"\", \"version\": \"v1\"}, \"owners\": [594797], \"revision\": 2, \"stats\": {\"downloads\": 1.0, \"unique_downloads\": 1.0, \"unique_views\": 10.0, \"version_downloads\": 1.0, \"version_unique_downloads\": 1.0, \"version_unique_views\": 10.0, \"version_views\": 12.0, \"version_volume\": 1672.0, \"views\": 12.0, \"volume\": 1672.0}, \"updated\": \"2023-08-09T02:26:49.823489+00:00\"}\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "print(json.dumps(hits[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k\n"
     ]
    }
   ],
   "source": [
    "if not hits: \n",
    "    print('k')\n",
    "else: \n",
    "    print('empty')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.DataFrame([]).empty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "notebooksearch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
